File: edu.wpi.first.wpilib.plugins.cpp/resources/templates/examples/PotentiometerPID/src/Robot.cpp
#include <array>

#include <AnalogInput.h>
#include <IterativeRobot.h>
#include <Joystick.h>
#include <PIDController.h>
#include <Spark.h>

/**
 * This is a sample program to demonstrate how to use a soft potentiometer and a
 * PID Controller to reach and maintain position setpoints on an elevator
 * mechanism.
 */
class Robot: public frc::IterativeRobot {
public:
	void RobotInit() override {
		pidController.SetInputRange(0, 5);
	}

	void TeleopInit() override {
		pidController.Enable();
	}

	void TeleopPeriodic() override {
		// when the button is pressed once, the selected elevator setpoint
		// is incremented
		bool currentButtonValue = joystick.GetTrigger();
		if (currentButtonValue && !previousButtonValue) {
			// index of the elevator setpoint wraps around.
			index = (index + 1) % (sizeof(kSetPoints) / 8);
		}
		previousButtonValue = currentButtonValue;

		pidController.SetSetpoint(kSetPoints[index]);
	}

private:
	static constexpr int kPotChannel = 1;
	static constexpr int kMotorChannel = 7;
	static constexpr int kJoystickChannel = 0;

	// Bottom, middle, and top elevator setpoints
	static constexpr std::array<double, 3> kSetPoints = { 1.0, 2.6, 4.3 };

	/* proportional, integral, and derivative speed constants; motor inverted
	 * DANGER: when tuning PID constants, high/inappropriate values for pGain,
	 * iGain, and dGain may cause dangerous, uncontrollable, or
	 * undesired behavior!
	 *
	 * These may need to be positive for a non-inverted motor
	 */
	static constexpr double kP = -5.0;
	static constexpr double kI = -0.02;
	static constexpr double kD = -2.0;

	int index = 0;
	bool previousButtonValue = false;

	frc::AnalogInput potentiometer { kPotChannel };
	frc::Joystick joystick { kJoystickChannel };
	frc::Spark elevatorMotor { kMotorChannel };

	/* potentiometer (AnalogInput) and elevatorMotor (Victor) can be used as a
	 * PIDSource and PIDOutput respectively. The PIDController takes pointers
	 * to the PIDSource and PIDOutput, so you must use &potentiometer and
	 * &elevatorMotor to get their pointers.
	 */
	frc::PIDController pidController { kP, kI, kD, &potentiometer,
			&elevatorMotor };
};

constexpr std::array<double, 3> Robot::kSetPoints;

START_ROBOT_CLASS(Robot)

File: edu.wpi.first.wpilib.plugins.java/resources/templates/examples/PotentiometerPID/src/org/usfirst/frc/team190/robot/Robot.java
package $package;

import edu.wpi.first.wpilibj.AnalogInput;
import edu.wpi.first.wpilibj.PIDController;
import edu.wpi.first.wpilibj.SampleRobot;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.Victor;

/**
 * This is a sample program to demonstrate how to use a soft potentiometer and a
 * PID controller to reach and maintain position setpoints on an elevator
 * mechanism.
 *
 * WARNING: While it may look like a good choice to use for your code if you're
 * inexperienced, don't. Unless you know what you are doing, complex code will
 * be much more difficult under this system. Use IterativeRobot or Command-Based
 * instead if you're new.
 */
public class Robot extends SampleRobot {
	final int kPotChannel = 1; // analog input pin
	final int kMotorChannel = 7; // PWM channel
	final int kJoystickChannel = 0; // usb number in DriverStation
	final int kButtonNumber = 4; // button on joystick

	final double kSetPoints[] = { 1.0, 2.6, 4.3 }; // bottom, middle, and top
													// elevator setpoints

	// proportional, integral, and derivative speed constants; motor inverted
	// DANGER: when tuning PID constants, high/inappropriate values for kP,
	// kI,
	// and kD may cause dangerous, uncontrollable, or undesired behavior!
	// these may need to be positive for a non-inverted motor
	final double kP = -5.0;
	final double kI = -0.02;
	final double kD = -2.0;

	PIDController pidController;
	AnalogInput potentiometer;
	Victor elevatorMotor;
	Joystick joystick;

	public Robot() {
		// make objects for potentiometer, the elevator motor controller, and
		// the joystick
		potentiometer = new AnalogInput(kPotChannel);
		elevatorMotor = new Victor(kMotorChannel);
		joystick = new Joystick(kJoystickChannel);

		// potentiometer (AnalogInput) and elevatorMotor (Victor) can be used as
		// a
		// PIDSource and PIDOutput respectively
		pidController = new PIDController(kP, kI, kD, potentiometer, elevatorMotor);
	}

	/**
	 * Runs during autonomous.
	 */
	@Override
	public void autonomous() {

	}

	/**
	 * Uses a PIDController and an array of setpoints to switch and maintain
	 * elevator positions. The elevator setpoint is selected by a joystick
	 * button.
	 */
	@Override
	public void operatorControl() {
		pidController.setInputRange(0, 5); // 0 to 5V
		pidController.setSetpoint(kSetPoints[0]); // set to first setpoint

		int index = 0;
		boolean currentValue;
		boolean previousValue = false;

		while (isOperatorControl() && isEnabled()) {
			pidController.enable(); // begin PID control

			// when the button is pressed once, the selected elevator setpoint
			// is incremented
			currentValue = joystick.getRawButton(kButtonNumber);
			if (currentValue && !previousValue) {
				pidController.setSetpoint(kSetPoints[index]);
				index = (index + 1) % kSetPoints.length; // index of elevator
														// setpoint wraps around
			}
			previousValue = currentValue;
		}
	}

	/**
	 * Runs during test mode.
	 */
	@Override
	public void test() {
	}
}
